radicle-cob/src/object/collaboration.rs

```diff
@@ -53,3 +53,43 @@ impl CollaborativeObject {
         self.history.tips().into_iter().map(Oid::from).collect()
     }
 }
+
+/// Takes a `refname` and performs a best attempt to extract out the
+/// [`TypeName`] and [`ObjectId`] from it.
+///
+/// This assumes that the `refname` is in a
+/// [`git_ref_format::Qualified`] format. If it has any
+/// `refs/namespaces`, they will be stripped to access the underlying
+/// [`git_ref_format::Qualified`] format.
+///
+/// In the [`git_ref_format::Qualified`] format it assumes that the
+/// reference name is of the form:
+///
+///   `refs/<category>/<typename>/<object_id>[/<rest>*]`
+///
+/// Note that their may be more components to the path after the
+/// [`ObjectId`] but they are ignored.
+///
+/// Also note that this will return `None` if:
+///
+///   * The `refname` is not [`git_ref_format::Qualified`]
+///   * The parsing of the [`ObjectId`] fails
+///   * The parsing of the [`TypeName`] fails
+pub fn parse_refstr<R>(name: &R) -> Option<(TypeName, ObjectId)>
+where
+    R: AsRef<git_ref_format::RefStr>,
+{
+    use git_ref_format::Qualified;
+    let name = name.as_ref();
+    let refs_cobs = match name.to_namespaced() {
+        None => Qualified::from_refstr(name)?,
+        Some(ns) => ns.strip_namespace_recursive(),
+    };
+
+    let (_refs, _cobs, typename, mut object_id) = refs_cobs.non_empty_components();
+    let object = object_id
+        .next()
+        .and_then(|oid| oid.parse::<ObjectId>().ok())?;
+    let name = typename.parse::<TypeName>().ok()?;
+    Some((name, object))
+}
```

Comment:
Maybe a small test to prove correctness?  It helps with review

Reply:
Yes, agree with this, especially beacuse I think there is a bug in the code now that I look at it ;)

Reply:
Well I've added a bunch of tests and no bugs showed up, so now I'm curious ;)

Reply:
:eyes: was just noticing that you're using `ObjectId` when parsing, instead of `git2::Oid`, which is what I had to change it to for it to work!

Reply:
Ya, the implementation for `FromStr` was wrong for `ObjectId` :) That's what I was pointing out when you were implementing the cobs. I guess I miscommunicated it :sweat_smile: